### PR TITLE
feat(FEC-10980): Enable setting ABR settings on-the-fly

### PR DESCRIPTION
**Pull request by Yuvalke**
### Description of the Changes

add an API applyABRRestriction to apply ABR restrictions and show only available tracks after restriction.

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/dash-adapter.js [hunk 1]
```diff
@@ -486,9 +487,35 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
...
    */
-  _maybeApplyAbrRestrictions(): void {
+  _maybeApplyAbrRestrictionsByBitrate(ABRConfig: Object): void {
+    if (ABRConfig.restrictions) {
+      const restrictions = ABRConfig.restrictions;
+      const minBitrate = restrictions.minBitrate ? restrictions.minBitrate : 0;
+      const maxBitrate = restrictions.maxBitrate ? restrictions.maxBitrate : Infinity;
+      if (maxBitrate < minBitrate) {
+        if (restrictions.minBitrate >= 0) {
+          this._config.capLevelToPlayerSize = false;
```
Comment (OrenMe) [inline src/dash-adapter.js:500]: this is by default false.
If an app set this and wants to apply other means of capping it needs to set it to false by itself.

File: src/dash-adapter.js [hunk 2]
```diff
@@ -221,6 +221,7 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
       }
       if (abr.restrictions) {
         if (abr.restrictions.minBitrate > 0) {
+          adapterConfig.capLevelToPlayerSize = false;
```
Comment (OrenMe) [inline src/dash-adapter.js:224]: Same here, this is by default false.
If an app set this and wants to apply other means of capping it needs to set it to false by itself.

File: src/dash-adapter.js [hunk 3]
```diff
@@ -486,9 +487,35 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
...
+      const restrictions = ABRConfig.restrictions;
+      const minBitrate = restrictions.minBitrate ? restrictions.minBitrate : 0;
+      const maxBitrate = restrictions.maxBitrate ? restrictions.maxBitrate : Infinity;
+      if (maxBitrate < minBitrate) {
+        if (restrictions.minBitrate >= 0) {
+          this._config.capLevelToPlayerSize = false;
+          this._config.shakaConfig.abr.restrictions.minBandwidth = restrictions.minBitrate;
+        }
+        if (restrictions.maxBitrate && restrictions.maxBitrate < Infinity) {
+          this._config.capLevelToPlayerSize = false;
```
Comment (OrenMe) [inline src/dash-adapter.js:504]: same here for `capLevelToPlayerSize`

**Review by OrenMe — CHANGES_REQUESTED** [review]

File: src/dash-adapter.js (hunk 2 above)
Comment (Yuvalke) [reply src/dash-adapter.js:224]: I did the same as we did for maxBitrate a few lines after I just wanted to keep the logic, once the restrictions set it'll change it to false.

File: src/dash-adapter.js [hunk 4]
```diff
@@ -471,50 +480,141 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
...
+  _updateRestriction(restrictions: PKABRRestrictionObject): void {
+    const shakaRestrictionsConfig = this._getRestrictionShakaConfig(restrictions);
+    this._shaka.configure({
+      abr: {
+        restrictions: shakaRestrictionsConfig
+      }
+    });
+  }
+
+  /**
```
Comment (Yuvalke) [inline src/dash-adapter.js:509]: remove unused code

File: src/dash-adapter.js [hunk 5]
```diff
@@ -471,50 +480,141 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
...
+        videoTracks.map(variant => variant[dim])
+      );
+    };
+    let restrictionsShakaConfig = {};
+    if (restrictions) {
+      let {maxHeight, maxWidth, maxBitrate, minHeight, minWidth, minBitrate} = restrictions;
+      if ((maxHeight || Infinity) >= (minHeight || 0)) {
+        const minVideoHeight = getMinDimensions('height');
+        restrictionsShakaConfig.minHeight = minHeight >= 0 ? minHeight : 0;
+        restrictionsShakaConfig.maxHeight = maxHeight >= minVideoHeight ? maxHeight : minVideoHeight;
```
Comment (Yuvalke) [inline src/dash-adapter.js:591]: add comment for this use case

File: src/dash-adapter.js [hunk 6]
```diff
@@ -220,13 +221,21 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
         adapterConfig.shakaConfig.abr.defaultBandwidthEstimate = abr.defaultBandwidthEstimate;
       }
       if (abr.restrictions) {
-        if (abr.restrictions.minBitrate > 0) {
-          adapterConfig.shakaConfig.abr.restrictions.minBandwidth = abr.restrictions.minBitrate;
-        }
-        if (abr.restrictions.maxBitrate < Infinity) {
-          //You can either set capping by size or bitrate, if bitrate is set then disable size capping
+        const {maxHeight, minHeight, maxWidth, minWidth, maxBitrate, minBitrate} = abr.restrictions;
```
Comment (Yuvalke) [inline src/dash-adapter.js:224]: remove will define with ABR changes on the fly

**Review by yairans — CHANGES_REQUESTED** [review]

**Review by yairans — DISMISSED** [review]

**Pull request merged**